File: java-kata/src/main/java/com/leetcode/BalancedBinaryTree.java
package com.leetcode;

import com.leetcode.data.TreeNode;

/**
 * <h3><a href="https://leetcode.com/problems/balanced-binary-tree/">Balanced Binary Tree</a></h3>
 * <p>
 *     Given a binary tree, determine if it is <strong>height-balanced</strong>.
 * </p>
 * <p>
 *     A <strong>height-balanced</strong> binary tree is a binary tree in which
 *     the depth of the two subtrees of every node never differs by more than one.
 * </p>
 */
public class BalancedBinaryTree {

    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        var leftHeight = balancedHeight(root.left);
        var rightHeight = balancedHeight(root.right);
        return leftHeight >= 0 && rightHeight >= 0 && Math.abs(leftHeight - rightHeight) <= 1;
    }

    private int balancedHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }

        var leftHeight = balancedHeight(root.left);
        var rightHeight = balancedHeight(root.right);
        var height = Math.max(Math.abs(leftHeight), Math.abs(rightHeight)) + 1;
        var isBalanced = leftHeight >= 0 && rightHeight >= 0 && Math.abs(leftHeight - rightHeight) <= 1;

        return isBalanced ? height : -height;
    }
}

File: kotlin-kata/src/main/kotlin/com/leetcode/BalancedBinaryTree.kt
package com.leetcode

import com.leetcode.data.TreeNode
import kotlin.math.absoluteValue
import kotlin.math.max

/**
 * ### [Balanced Binary Tree](https://leetcode.com/problems/balanced-binary-tree/)
 *
 * Given a binary tree, determine if it is **height-balanced**.
 *
 * A **height-balanced** binary tree is a binary tree in which
 * the depth of the two subtrees of every node never differs by more than one.
 */
class BalancedBinaryTree {
    fun isBalanced(root: TreeNode?): Boolean {
        if (root == null) {
            return true
        }

        return isBalanced(balancedAwareHeight(root.left), balancedAwareHeight(root.right))
    }

    // Positive - balanced, negative - no
    private fun balancedAwareHeight(node: TreeNode?): Int {
        if (node == null) {
            return 0
        }

        var leftHeight = balancedAwareHeight(node.left)
        var rightHeight = balancedAwareHeight(node.right)
        var height = max(leftHeight.absoluteValue, rightHeight.absoluteValue) + 1

        return if (isBalanced(leftHeight, rightHeight)) height else -height
    }

    private inline fun isBalanced(leftHeight: Int, rightHeight: Int): Boolean {
        return leftHeight >= 0 && rightHeight >= 0 && (leftHeight - rightHeight).absoluteValue <= 1
    }
}
